server/middleware/applicationAuthMiddleware.ts

```diff
@@ -1,12 +1,16 @@
 import type { NextFunction, Request, RequestHandler, Response } from 'express'
 import { MiddlewareSpec } from '../@types/ui'
-import { ApprovedPremisesUserRole } from '../@types/shared'
+import { ApprovedPremisesUserPermission, ApprovedPremisesUserRole } from '../@types/shared'
 import logger from '../../logger'
+import { hasPermission } from '../utils/users'
 
 export default function applicationAuthMiddleware(requestHandler: RequestHandler, middlewareSpec?: MiddlewareSpec) {
-  if (middlewareSpec?.allowedRoles) {
+  if (middlewareSpec?.allowedRoles || middlewareSpec?.allowedPermissions) {
     return async (req: Request, res: Response, next: NextFunction) => {
-      if (res.locals.user.roles.some((role: ApprovedPremisesUserRole) => middlewareSpec.allowedRoles.includes(role))) {
+      if (
+        res.locals.user.roles.some((role: ApprovedPremisesUserRole) => middlewareSpec.allowedRoles.includes(role)) ||
+        hasPermission(res.locals.user, middlewareSpec.allowedPermissions ?? new Array<ApprovedPremisesUserPermission>())
```

Comment:
This isn't a blocking comment as I know that the function wasn't added in this PR but it seems a little off to me that `hasPermission` takes an array of Permissions but it's name implies that it tests only one permission. It is also only ever called with an array containing 1 item. This difference in its signature with `hasRole` which makes this `if` harder to read than it needs to be.  
Separately, could remove  `?? new Array<ApprovedPremisesUserPermission>()` it isn't necessary as we check that `middlewareSpec.allowedPermissions` is truthy on line 8, therefore (according to its type) it must be an array.

Reply:
Taken action to rename `hasPermission` to hasPermissions in a separate PR.
Taken action to improve readability of code to cover role check as well as permissions.